File: verifier/js/verify.mjs
// Reference verifier for the randomness beacon.
//
// WebCrypto only. No dependencies, nothing to install.
//
//	node verify.mjs emission.jws
//	node verify.mjs https://.../2026/08/29/1730-emission.jws <public-key-hex>
//
// The core — derive(), deriveBlock(), Stream — runs unchanged in a browser.
// Only readSource() is Node-specific.

const HKDF_BLOCK = 64;
const DRAND_RELAY = "https://drand.cloudflare.com";

const subtle = globalThis.crypto?.subtle ?? (await import("node:crypto")).webcrypto.subtle;

// --- The byte stream -------------------------------------------------------

class Stream {
  constructor(seed, publicValue, context) {
    this.seed = seed;
    this.public = publicValue;
    this.context = context;
    this.buf = new Uint8Array(0);
    this.pos = 0;
  }

  async nextByte() {
    if (this.pos >= this.buf.length) {
      const info = concat(this.context, encode("|" + this.buf.length));
      const block = await hkdf(this.seed, this.public, info, HKDF_BLOCK);
      this.buf = concat(this.buf, block);
    }
    return this.buf[this.pos++];
  }

  // Uniform value in [0,limit) with no modulo bias: values above the last
  // whole multiple of limit are discarded and drawn again.
  async integer(limit) {
    if (limit <= 1) return 0;

    let width = 1;
    let space = 256;
    while (space < limit) {
      width++;
      space *= 256;
    }
    const cut = Math.floor(space / limit) * limit;

    for (;;) {
      let v = 0;
      for (let i = 0; i < width; i++) v = v * 256 + (await this.nextByte());
      if (v < cut) return v % limit;
    }
  }
}

// hkdf is HKDF-SHA256, RFC 5869. WebCrypto does both stages in one call.
async function hkdf(ikm, salt, info, length) {
  const key = await subtle.importKey("raw", ikm, "HKDF", false, ["deriveBits"]);
  const bits = await subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt, info },
    key,
    length * 8,
  );
  return new Uint8Array(bits);
}

// --- Derivation ------------------------------------------------------------

// A variant is published in one of two shapes, and this is the first thing a
// reimplementation gets wrong:
//
//   - composite carries an explicit "blocks" array
//   - every other group carries its parameters FLAT, and the mold is the
//     GROUP NAME, not a field
//
// Two more traps in the flat shape: random_matrix carries rows/cols instead of
// count, and truncated_continuous publishes min/max already multiplied by
// scale.
function blocksOf(group, v) {
  if (group === "composite") return v.blocks;
  const count = group === "random_matrix" ? v.rows * v.cols : v.count;
  return [{ mold: group, count, min: v.min, max: v.max }];
}

async function derive(em, name, spec) {
  const seed = fromHex(em.seed);
  const publicValue = fromHex(em.round.randomness);
  const base = `${em.version}|${name}|${em.drand_round}`;

  const blocks = [];
  for (let i = 0; i < spec.length; i++) {
    const s = new Stream(seed, publicValue, encode(`${base}|b${i}`));
    blocks.push(await deriveBlock(s, spec[i]));
  }
  return shape(name, spec, blocks);
}

async function deriveBlock(s, b) {
  const universe = b.max - b.min + 1;
  if (universe < 1) throw new Error(`empty universe [${b.min}..${b.max}]`);
  const count = b.count || 1;

  switch (b.mold) {
    case "discrete_uniform":
      return [b.min + (await s.integer(universe))];

    case "sample_without_replacement": {
      if (count > universe) throw new Error(`asks for ${count} out of ${universe}`);
      const out = await sampleWithoutReplacement(s, universe, count, b.min);
      return out.sort((x, y) => x - y);
    }

    case "permutation":
      if (count > universe) throw new Error(`asks for ${count} out of ${universe}`);
      return sampleWithoutReplacement(s, universe, count, b.min);

    case "uniform_vector_with_replacement":
    case "random_matrix": {
      const out = [];
      for (let i = 0; i < count; i++) out.push(b.min + (await s.integer(universe)));
      return out;
    }

    // min and max already carry the scale, so the span is read straight off
    // the file. Multiplying again is the classic reimplementation bug.
    case "truncated_continuous":
      return [b.min + (await s.integer(b.max - b.min + 1))];
  }
  throw new Error(`unknown mold "${b.mold}"`);
}

// Draws k distinct values in the order they come out, with a partial
// Fisher-Yates over a sparse map.
async function sampleWithoutReplacement(s, universe, count, min) {
  const moved = new Map();
  const out = [];

  for (let i = 0; i < count; i++) {
    const j = i + (await s.integer(universe - i));
    const vj = moved.has(j) ? moved.get(j) : j;
    const vi = moved.has(i) ? moved.get(i) : i;
    moved.set(j, vi);
    moved.set(i, vj);
    out.push(min + vj);
  }
  return out;
}

// shape renders the result the way the file publishes it. The shape depends on
// the mold: a scalar, a list, a list of rows, or one list per block.
function shape(name, spec, values) {
  if (spec.length > 1) return values;
  const flat = values[0];

  switch (spec[0].mold) {
    case "discrete_uniform":
    case "truncated_continuous":
      return flat[0];

    case "random_matrix": {
      const [rows, cols] = dimensions(name, flat.length);
      const out = [];
      for (let i = 0; i < rows; i++) out.push(flat.slice(i * cols, (i + 1) * cols));
      return out;
    }
  }
  return flat;
}

// dimensions reads the matrix shape from the variant name, m-<rows>x<cols>.
function dimensions(name, total) {
  const m = /^m-(\d+)x(\d+)$/.exec(name);
  if (m) {
    const rows = Number(m[1]);
    const cols = Number(m[2]);
    if (rows > 0 && cols > 0 && rows * cols === total) return [rows, cols];
  }
  return [1, total];
}

// --- Checks ----------------------------------------------------------------

async function main() {
  const [source, keyHex] = process.argv.slice(2);
  if (!source) {
    console.log("usage: node verify.mjs <file.jws|url> [public-key-hex]");
    process.exit(2);
  }

  let raw;
  try {
    raw = (await readSource(source)).trim();
  } catch (e) {
    unverifiable(`cannot read the file: ${e.message}`);
  }

  const parts = raw.split(".");
  if (parts.length !== 3) notVerified(`expected 3 dot-separated parts, got ${parts.length}`);
  const header = JSON.parse(decode(fromB64Url(parts[0])));
  const em = JSON.parse(decode(fromB64Url(parts[1])));

  console.log(
    `file      : ${em.type} turn ${em.utc_turn}, version ${em.version}, drand round ${em.drand_round}`,
  );
  console.log(`key id    : ${header.kid}`);

  // 1. Signature, over the RAW text: re-serializing the JSON changes the bytes
  // and breaks a signature that is perfectly valid.
  if (keyHex) {
    const signed = encode(parts[0] + "." + parts[1]);
    let ok;
    try {
      const key = await subtle.importKey("raw", fromHex(keyHex), "Ed25519", false, ["verify"]);
      ok = await subtle.verify("Ed25519", key, fromB64Url(parts[2]), signed);
    } catch {
      unverifiable("this runtime has no Ed25519 in WebCrypto");
    }
    if (!ok) notVerified("the Ed25519 signature does not verify");
    console.log("signature : OK");
  } else {
    console.log("signature : SKIPPED, no public key given");
  }

  // 2. The revealed seed must hash to what was committed.
  const sum = new Uint8Array(await subtle.digest("SHA-256", fromHex(em.seed)));
  if (toHex(sum) !== em.seed_sha256) {
    notVerified("SHA-256 of the seed does not match seed_sha256");
  }
  console.log("seed      : OK, matches its own hash");

  // 3. The public value comes from drand, not from us. Ask them.
  try {
    const body = await readSource(`${DRAND_RELAY}/public/${em.drand_round}`);
    if (JSON.parse(body).randomness !== em.round.randomness) {
      notVerified(`drand round ${em.drand_round} carries a different randomness`);
    }
    console.log(`drand     : OK, round ${em.drand_round} carries this randomness`);
  } catch (e) {
    console.log(`drand     : SKIPPED, ${e.message}`);
  }

  // 4. Every number, reproduced from the seed and the round.
  let checked = 0;
  let failed = 0;
  for (const [group, variants] of Object.entries(em.variants)) {
    for (const [name, v] of Object.entries(variants)) {
      const got = JSON.stringify(await derive(em, name, blocksOf(group, v)));
      const want = JSON.stringify(v.result);
      if (got !== want) {
        failed++;
        console.log(`  ${name.padEnd(24)} MISMATCH\n     expected ${want}\n     computed ${got}`);
        continue;
      }
      checked++;
    }
  }
  if (failed > 0) notVerified(`${failed} of ${checked + failed} variants do not reproduce`);
  console.log(`numbers   : OK, ${checked} variants reproduced`);

  console.log("\nVERIFIED");
}

// --- Plumbing --------------------------------------------------------------

const encode = (s) => new TextEncoder().encode(s);
const decode = (b) => new TextDecoder().decode(b);

function concat(a, b) {
  const out = new Uint8Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

function fromHex(hex) {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(hex.substr(i * 2, 2), 16);
  return out;
}

const toHex = (bytes) => [...bytes].map((b) => b.toString(16).padStart(2, "0")).join("");

function fromB64Url(s) {
  const b64 = s.replace(/-/g, "+").replace(/_/g, "/");
  const bin = atob(b64.padEnd(b64.length + ((4 - (b64.length % 4)) % 4), "="));
  return Uint8Array.from(bin, (c) => c.charCodeAt(0));
}

async function readSource(source) {
  if (/^https?:\/\//.test(source)) {
    const resp = await fetch(source);
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    return resp.text();
  }
  const { readFile } = await import("node:fs/promises");
  return readFile(source, "utf8");
}

function notVerified(message) {
  console.log(`\nNOT VERIFIED: ${message}`);
  process.exit(1);
}

function unverifiable(message) {
  console.log(`\nUNVERIFIABLE: ${message}`);
  process.exit(3);
}

await main();

File: verifier/py/verify.py
#!/usr/bin/env python3
"""Reference verifier for the randomness beacon.

    python3 verify.py emission.jws
    python3 verify.py https://.../2026/08/29/1730-emission.jws <public-key-hex>

Hashing and derivation use the standard library only.

The Ed25519 signature needs one dependency:

    cryptography    pip install cryptography
                    https://pypi.org/project/cryptography/
                    Python Cryptographic Authority, backed by OpenSSL

Without it the verifier says UNVERIFIABLE for that check instead of guessing.
"""

import hashlib
import hmac
import json
import sys
import urllib.request
from base64 import urlsafe_b64decode

HKDF_BLOCK = 64
DRAND_RELAY = "https://drand.cloudflare.com"
HTTP_TIMEOUT = 20


# --- The byte stream -------------------------------------------------------


class Stream:
    def __init__(self, seed, public, context):
        self.seed = seed
        self.public = public
        self.context = context
        self.buf = b""
        self.pos = 0

    def next_byte(self):
        if self.pos >= len(self.buf):
            info = self.context + b"|" + str(len(self.buf)).encode()
            self.buf += hkdf(self.seed, self.public, info, HKDF_BLOCK)
        b = self.buf[self.pos]
        self.pos += 1
        return b

    def integer(self, limit):
        """Uniform value in [0,limit) with no modulo bias: values above the
        last whole multiple of limit are discarded and drawn again."""
        if limit <= 1:
            return 0

        width, space = 1, 256
        while space < limit:
            width += 1
            space *= 256
        cut = (space // limit) * limit

        while True:
            v = 0
            for _ in range(width):
                v = v * 256 + self.next_byte()
            if v < cut:
                return v % limit


def hkdf(ikm, salt, info, length):
    """HKDF-SHA256, RFC 5869, both stages."""
    prk = hmac.new(salt, ikm, hashlib.sha256).digest()

    out, block, i = b"", b"", 1
    while len(out) < length:
        block = hmac.new(prk, block + info + bytes([i]), hashlib.sha256).digest()
        out += block
        i += 1
    return out[:length]


# --- Derivation ------------------------------------------------------------


def blocks_of(group, v):
    """A variant is published in one of two shapes, and this is the first thing
    a reimplementation gets wrong:

      - composite carries an explicit "blocks" array
      - every other group carries its parameters FLAT, and the mold is the
        GROUP NAME, not a field

    Two more traps in the flat shape: random_matrix carries rows/cols instead
    of count, and truncated_continuous publishes min/max already multiplied by
    scale.
    """
    if group == "composite":
        return v["blocks"]
    count = v["rows"] * v["cols"] if group == "random_matrix" else v.get("count", 1)
    return [{"mold": group, "count": count, "min": v["min"], "max": v["max"]}]


def derive(em, name, spec):
    seed = bytes.fromhex(em["seed"])
    public = bytes.fromhex(em["round"]["randomness"])
    base = f"{em['version']}|{name}|{em['drand_round']}"

    blocks = []
    for i, b in enumerate(spec):
        s = Stream(seed, public, f"{base}|b{i}".encode())
        blocks.append(derive_block(s, b))
    return shape(name, spec, blocks)


def derive_block(s, b):
    universe = b["max"] - b["min"] + 1
    if universe < 1:
        raise ValueError(f"empty universe [{b['min']}..{b['max']}]")
    count = b.get("count") or 1
    mold = b["mold"]

    if mold == "discrete_uniform":
        return [b["min"] + s.integer(universe)]

    if mold == "sample_without_replacement":
        if count > universe:
            raise ValueError(f"asks for {count} out of {universe}")
        return sorted(sample_without_replacement(s, universe, count, b["min"]))

    if mold == "permutation":
        if count > universe:
            raise ValueError(f"asks for {count} out of {universe}")
        return sample_without_replacement(s, universe, count, b["min"])

    if mold in ("uniform_vector_with_replacement", "random_matrix"):
        return [b["min"] + s.integer(universe) for _ in range(count)]

    # min and max already carry the scale, so the span is read straight off the
    # file. Multiplying again is the classic reimplementation bug.
    if mold == "truncated_continuous":
        return [b["min"] + s.integer(b["max"] - b["min"] + 1)]

    raise ValueError(f'unknown mold "{mold}"')


def sample_without_replacement(s, universe, count, minimum):
    """Draws k distinct values in the order they come out, with a partial
    Fisher-Yates over a sparse map."""
    moved, out = {}, []

    for i in range(count):
        j = i + s.integer(universe - i)
        vj = moved.get(j, j)
        vi = moved.get(i, i)
        moved[j], moved[i] = vi, vj
        out.append(minimum + vj)
    return out


def shape(name, spec, values):
    """Renders the result the way the file publishes it: a scalar, a list, a
    list of rows, or one list per block."""
    if len(spec) > 1:
        return values
    flat = values[0]
    mold = spec[0]["mold"]

    if mold in ("discrete_uniform", "truncated_continuous"):
        return flat[0]

    if mold == "random_matrix":
        rows, cols = dimensions(name, len(flat))
        return [flat[i * cols : (i + 1) * cols] for i in range(rows)]

    return flat


def dimensions(name, total):
    """Reads the matrix shape from the variant name, m-<rows>x<cols>."""
    import re

    m = re.fullmatch(r"m-(\d+)x(\d+)", name)
    if m:
        rows, cols = int(m.group(1)), int(m.group(2))
        if rows > 0 and cols > 0 and rows * cols == total:
            return rows, cols
    return 1, total


# --- Checks ----------------------------------------------------------------


def main():
    if len(sys.argv) < 2:
        print("usage: python3 verify.py <file.jws|url> [public-key-hex]")
        sys.exit(2)
    source = sys.argv[1]
    key_hex = sys.argv[2] if len(sys.argv) > 2 else None

    try:
        raw = read_source(source).strip()
    except Exception as e:
        unverifiable(f"cannot read the file: {e}")

    parts = raw.split(".")
    if len(parts) != 3:
        not_verified(f"expected 3 dot-separated parts, got {len(parts)}")
    header = json.loads(b64(parts[0]))
    em = json.loads(b64(parts[1]))

    print(
        f"file      : {em['type']} turn {em['utc_turn']}, "
        f"version {em['version']}, drand round {em['drand_round']}"
    )
    print(f"key id    : {header.get('kid')}")

    # 1. Signature, over the RAW text: re-serializing the JSON changes the
    # bytes and breaks a signature that is perfectly valid.
    if key_hex:
        try:
            from cryptography.exceptions import InvalidSignature
            from cryptography.hazmat.primitives.asymmetric.ed25519 import (
                Ed25519PublicKey,
            )
        except ImportError:
            unverifiable(
                "Ed25519 needs `cryptography`: pip install cryptography  "
                "(https://pypi.org/project/cryptography/)"
            )

        key = Ed25519PublicKey.from_public_bytes(bytes.fromhex(key_hex.strip()))
        signed = (parts[0] + "." + parts[1]).encode()
        try:
            key.verify(b64(parts[2]), signed)
        except InvalidSignature:
            not_verified("the Ed25519 signature does not verify")
        print("signature : OK")
    else:
        print("signature : SKIPPED, no public key given")

    # 2. The revealed seed must hash to what was committed.
    if hashlib.sha256(bytes.fromhex(em["seed"])).hexdigest() != em["seed_sha256"]:
        not_verified("SHA-256 of the seed does not match seed_sha256")
    print("seed      : OK, matches its own hash")

    # 3. The public value comes from drand, not from us. Ask them.
    try:
        body = read_source(f"{DRAND_RELAY}/public/{em['drand_round']}")
        if json.loads(body)["randomness"] != em["round"]["randomness"]:
            not_verified(f"drand round {em['drand_round']} carries a different randomness")
        print(f"drand     : OK, round {em['drand_round']} carries this randomness")
    except SystemExit:
        raise
    except Exception as e:
        print(f"drand     : SKIPPED, {e}")

    # 4. Every number, reproduced from the seed and the round.
    checked = failed = 0
    for group, variants in em["variants"].items():
        for name, v in variants.items():
            got = json.dumps(derive(em, name, blocks_of(group, v)), separators=(",", ":"))
            want = json.dumps(v["result"], separators=(",", ":"))
            if got != want:
                failed += 1
                print(f"  {name:<24} MISMATCH\n     expected {want}\n     computed {got}")
                continue
            checked += 1

    if failed:
        not_verified(f"{failed} of {checked + failed} variants do not reproduce")
    print(f"numbers   : OK, {checked} variants reproduced")

    print("\nVERIFIED")


# --- Plumbing --------------------------------------------------------------


def b64(s):
    return urlsafe_b64decode(s + "=" * (-len(s) % 4))


def read_source(source):
    if source.startswith(("http://", "https://")):
        # A User-Agent is required: the default one urllib sends is rejected
        # by some CDNs with a 403, drand's included.
        req = urllib.request.Request(source, headers={"User-Agent": "beacon-verifier"})
        with urllib.request.urlopen(req, timeout=HTTP_TIMEOUT) as r:
            return r.read().decode()
    with open(source, encoding="utf-8") as fh:
        return fh.read()


def not_verified(message):
    print(f"\nNOT VERIFIED: {message}")
    sys.exit(1)


def unverifiable(message):
    print(f"\nUNVERIFIABLE: {message}")
    sys.exit(3)


if __name__ == "__main__":
    main()
